src/bun.js/bindings/JSCommonJSExtensions.cpp

```diff
@@ -0,0 +1,181 @@
+#include "JSCommonJSExtensions.h"
+#include "ZigGlobalObject.h"
+
+namespace Bun {
+using namespace JSC;
+
+const JSC::ClassInfo JSCommonJSExtensions::s_info = { "CommonJSExtensions"_s, &Base::s_info, nullptr, nullptr, CREATE_METHOD_TABLE(JSCommonJSExtensions) };
+
+// These functions are implemented as no-ops because it doesn't seem like any
+// projects call them directly. They are defined separately so that assigning
+// one to the other can be detected and use the corresponding loader.
+JSC_DEFINE_HOST_FUNCTION(jsLoaderJS, (JSC::JSGlobalObject * globalObject, JSC::CallFrame* callFrame))
+{
+    auto scope = DECLARE_THROW_SCOPE(globalObject->vm());
+    throwTypeError(globalObject, scope, "Calling Module._extensions[\".js\"] directly is not implemented."_s);
+    return JSValue::encode({});
+}
+JSC_DEFINE_HOST_FUNCTION(jsLoaderJSON, (JSC::JSGlobalObject * globalObject, JSC::CallFrame* callFrame))
+{
+    auto scope = DECLARE_THROW_SCOPE(globalObject->vm());
+    throwTypeError(globalObject, scope, "Calling Module._extensions[\".json\"] directly is not implemented."_s);
+    return JSValue::encode({});
+}
+JSC_DEFINE_HOST_FUNCTION(jsLoaderNode, (JSC::JSGlobalObject * globalObject, JSC::CallFrame* callFrame))
+{
+    auto scope = DECLARE_THROW_SCOPE(globalObject->vm());
+    throwTypeError(globalObject, scope, "Calling Module._extensions[\".node\"] directly is not implemented."_s);
+    return JSValue::encode({});
+}
+JSC_DEFINE_HOST_FUNCTION(jsLoaderTS, (JSC::JSGlobalObject * globalObject, JSC::CallFrame* callFrame))
+{
+    auto scope = DECLARE_THROW_SCOPE(globalObject->vm());
+    throwTypeError(globalObject, scope, "Calling Module._extensions[\".ts\"] directly is not implemented."_s);
+    return JSValue::encode({});
+}
+
+void JSCommonJSExtensions::finishCreation(JSC::VM& vm)
+{
+    Base::finishCreation(vm);
+    ASSERT(inherits(info()));
+
+    Zig::GlobalObject* global = defaultGlobalObject(globalObject());
+    JSC::JSFunction* fnLoadJS = JSC::JSFunction::create(
```

Comment:
Can these be cached to avoid redundant JSFunction creation?

Reply:
this is called once?